File: src/platform/decode.cpp
/**
   \file decode.cpp
   \copyright 2014 Jesse Haber-Kucharsky

   \note Released under the terms of the Apache License Version 2.0. See
   \p /nebula/LICENSE or http://www.apache.org/licenses/
 */

#include "decode.hpp"

#include <array>

namespace nebula {

const static std::array<Register, 8> kRegisters{
    {Register::A, Register::B, Register::C, Register::X,
     Register::Y, Register::Z, Register::I, Register::J}};

static optional<Register> decodeRegister(Word code) {
  if (code >= kRegisters.size()) {
    return {};
  }

  return kRegisters[static_cast<std::size_t>(code)];
}

optional<OpCode> decodeOpCode(Word code) {
  switch (code) {
    case 0x01u:
      return OpCode::Set;
    case 0x02u:
      return OpCode::Add;
    case 0x03u:
      return OpCode::Sub;
    case 0x04u:
      return OpCode::Mul;
    case 0x05u:
      return OpCode::Mli;
    case 0x06u:
      return OpCode::Div;
    case 0x07u:
      return OpCode::Dvi;
    case 0x08u:
      return OpCode::Mod;
    case 0x09u:
      return OpCode::Mdi;
    case 0x0au:
      return OpCode::And;
    case 0x0bu:
      return OpCode::Bor;
    case 0x0cu:
      return OpCode::Xor;
    case 0x0du:
      return OpCode::Shr;
    case 0x0eu:
      return OpCode::Asr;
    case 0x0fu:
      return OpCode::Shl;
    case 0x10u:
      return OpCode::Ifb;
    case 0x11u:
      return OpCode::Ifc;
    case 0x12u:
      return OpCode::Ife;
    case 0x13u:
      return OpCode::Ifn;
    case 0x14u:
      return OpCode::Ifg;
    case 0x15u:
      return OpCode::Ifa;
    case 0x16u:
      return OpCode::Ifl;
    case 0x17u:
      return OpCode::Ifu;
    case 0x1au:
      return OpCode::Adx;
    case 0x1bu:
      return OpCode::Sbx;
    case 0x1eu:
      return OpCode::Sti;
    case 0x1fu:
      return OpCode::Std;
  }

  return {};
}

optional<SpecialOpCode> decodeSpecialOpCode(Word code) {
  switch (code) {
    case 0x01u:
      return SpecialOpCode::Jsr;
    case 0x08u:
      return SpecialOpCode::Int;
    case 0x09u:
      return SpecialOpCode::Iag;
    case 0x0au:
      return SpecialOpCode::Ias;
    case 0x0bu:
      return SpecialOpCode::Rfi;
    case 0x0cu:
      return SpecialOpCode::Iaq;
    case 0x10u:
      return SpecialOpCode::Hwn;
    case 0x11u:
      return SpecialOpCode::Hwq;
    case 0x12u:
      return SpecialOpCode::Hwi;
    case 0x15u:
      return SpecialOpCode::Abt;
  }

  return {};
}

unique_ptr<IAddress> decodeAddress(AddressContext context, Word code) {
  if (code <= 0x07u) {
    return make_unique<address::RegisterDirect>(*decodeRegister(code));
  } else if ((code >= 0x08u) && (code <= 0x0fu)) {
    return make_unique<address::RegisterIndirect>(
        *decodeRegister(code - 0x08u));
  } else if ((code >= 0x10u) && (code <= 0x17u)) {
    return make_unique<address::RegisterIndirectOffset>(
        *decodeRegister(code - 0x10u));
  } else if ((context == AddressContext::B) && (code == 0x18u)) {
    return make_unique<address::Push>();
  } else if ((context == AddressContext::A) && (code == 0x18u)) {
    return make_unique<address::Pop>();
  } else if (code == 0x19u) {
    return make_unique<address::Peek>();
  } else if (code == 0x1au) {
    return make_unique<address::Pick>();
  } else if (code == 0x1bu) {
    return make_unique<address::Sp>();
  } else if (code == 0x1cu) {
    return make_unique<address::Pc>();
  } else if (code == 0x1du) {
    return make_unique<address::Ex>();
  } else if (code == 0x1eu) {
    return make_unique<address::Indirect>();
  } else if (code == 0x1fu) {
    return make_unique<address::Direct>();
  } else if ((context == AddressContext::A) && (code >= 0x20u) &&
             (code <= 0x3fu)) {
    return make_unique<address::FastDirect>(code - 0x21u);
  } else {
    return nullptr;
  }
}

static unique_ptr<IInstruction> decodeBinaryInstruction(Word word) {
  // Is there a nicer way to do this?

  auto op_code = decodeOpCode(word & 0x1fu);
  if (!op_code) return nullptr;

  auto address_a = decodeAddress(AddressContext::A, (word & 0xfc00u) >> 10u);
  if (!address_a) return nullptr;

  auto address_b = decodeAddress(AddressContext::B, (word & 0x3e0) >> 5u);
  if (!address_b) return nullptr;

  return make_unique<instruction::Binary>(
      *op_code, std::move(address_b), std::move(address_a));
}

static unique_ptr<IInstruction> decodeUnaryInstruction(Word word) {
  auto op_code = decodeSpecialOpCode((word & 0x3e0) >> 5u);
  if (!op_code) return nullptr;

  auto address = decodeAddress(AddressContext::A, (word & 0xfc00) >> 10u);
  if (!address) return nullptr;

  return make_unique<instruction::Unary>(*op_code, std::move(address));
}

unique_ptr<IInstruction> decodeInstruction(Word word) {
  auto binary_instruction = decodeBinaryInstruction(word);
  if (binary_instruction) {
    return binary_instruction;
  } else {
    return decodeUnaryInstruction(word);
  }
}

}  // namespace nebula

File: src/emu/decode.ml
(** @author Jesse Haber-Kucharsky
    @see 'LICENSE' License details *)

open Common

open Functional

module F = Option.Functor

type context = A | B

let code w =
  let open Code in
  match Word.to_int w with
  | 0x01 -> Some Set
  | 0x02 -> Some Add
  | 0x03 -> Some Sub
  | 0x04 -> Some Mul
  | 0x05 -> Some Mli
  | 0x06 -> Some Div
  | 0x07 -> Some Dvi
  | 0x0b -> Some Bor
  | 0x0f -> Some Shl
  | 0x12 -> Some Ife
  | 0x13 -> Some Ifn
  | 0x14 -> Some Ifg
  | 0x16 -> Some Ifl
  | 0x1e -> Some Sti
  | _ -> None

let special_code w =
  let open Special_code in
  match Word.to_int w with
  | 0x01 -> Some Jsr
  | 0x08 -> Some Int
  | 0x09 -> Some Iag
  | 0x0a -> Some Ias
  | 0x0b -> Some Rfi
  | 0x0c -> Some Iaq
  | 0x10 -> Some Hwn
  | 0x11 -> Some Hwq
  | 0x12 -> Some Hwi
  | 0x1e -> Some Abt
  | 0x1f -> Some Dbg
  | _ -> None

let register w =
  match Word.to_int w with
  | 0 -> Some Reg.A
  | 1 -> Some Reg.B
  | 2 -> Some Reg.C
  | 3 -> Some Reg.X
  | 4 -> Some Reg.Y
  | 5 -> Some Reg.Z
  | 6 -> Some Reg.I
  | 7 -> Some Reg.J
  | _ -> None

let address ctx w =
  let i = Word.to_int w in

  if i <= 0x07 then
    register w |> F.map (fun r -> Address.Reg_direct r)
  else if (i >= 0x08) && (i < 0x0f) then
    register Word.(w - word 0x08) |> F.map (fun r -> Address.Reg_indirect r)
  else if (i >= 0x10) && (i <= 0x17) then
    register Word.(w - word 0x10) |> F.map (fun r -> Address.Reg_indirect_offset r)
  else match i with
    | 0x18 when ctx = A -> Some Address.Pop
    | 0x18 when ctx = B -> Some Address.Push
    | 0x19 -> Some Address.Peek
    | 0x1a -> Some Address.Pick
    | 0x1b -> Some Address.SP
    | 0x1c -> Some Address.PC
    | 0x1d -> Some Address.EX
    | 0x1e -> Some Address.Indirect
    | 0x1f -> Some Address.Direct
    | _ ->
      if (ctx = A) && (i >= 0x20) && (i <= 0x3f) then
        Some (Address.Literal Word.(w - word 0x21))
      else
        None

let instruction w =
  let open Option.Monad in

  let binary w =
    code Word.(w land word 0x1f) >>= fun c ->
    address A Word.((w land word 0xfc00) lsr 10) >>= fun a ->
    address B Word.((w land word 0x3e0) lsr 5) >>= fun b ->
    Some (Instruction.Binary (c, b, a))
  in

  let unary w =
    special_code Word.((w land word 0x3e0) lsr 5) >>= fun c ->
    address A Word.((w land word 0xfc00) lsr 10) >>= fun a ->
    Some (Instruction.Unary (c, a))
  in

  match binary w with
  | Some ins -> Some ins
  | None -> unary w
